core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -467,7 +467,9 @@ private[spark] class Master(
    * two executors on the same worker).
    */
   def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+    worker.memoryFree >= app.desc.memoryPerSlave &&
+    !worker.hasExecutor(app) &&
+    !app.removedExecutors.exists(_.worker == worker)
```

Comment:
I agree with @mateiz, and Spark Streaming is by no means the only Spark application for which a common use case is to have the entire cluster essentially running a single long-lived application.  To have transiently failing Executors perpetually banned from doing work for that long-lived application is not acceptable.